dart/dynamics/VoxelGridShape.cpp

```diff
@@ -107,16 +131,43 @@ fcl_shared_ptr<const octomap::OcTree> VoxelGridShape::getOctree() const
 
 //==============================================================================
 void VoxelGridShape::updateOccupancy(
-    const octomap::Pointcloud& pointCloud, const octomap::point3d& sensorOrigin)
+    const Eigen::Vector3d& point, bool occupied)
 {
-  mOctree->insertPointCloud(pointCloud, sensorOrigin);
+  mOctree->updateNode(toVector3(point), occupied);
 }
 
 //==============================================================================
 void VoxelGridShape::updateOccupancy(
-    const Eigen::Vector3d& point, bool occupied)
+    const Eigen::Vector3d& from, const Eigen::Vector3d& to)
+{
+  mOctree->insertRay(toVector3(from), toVector3(to));
+}
+
+//==============================================================================
+void VoxelGridShape::updateOccupancy(
+    const octomap::Pointcloud& pointCloud,
+    const Eigen::Vector3d& sensorOrigin,
+    const Frame* inCoordinatesOf)
+{
+  if (inCoordinatesOf == Frame::World())
+  {
+    mOctree->insertPointCloud(pointCloud, toVector3(sensorOrigin));
```

Comment:
> Just to be clear, will this also clear out the previous occupancy? To me, the function name insertPointCloud implies that it will add these points to the already existing occupancy grid, ...

You're correct. More precisely, the function updates the occupancy probabilities. The probability at the end point will be increased, and the probabilities at the points that the rays pass through will be decreased. I used "update" term in that sense. I just couldn't come up with better name for that. Any suggestions? :smile:

Reply:
I think I follow. Does this function have hysteresis, then? Where the previous state will have an effect on the current state? If so, it might be good to mention that in the documentation of `updateOccupancy`.

I think the word "update" is reasonable, or at least I don't have a better suggestion off the top of my head. But I think sometimes "update" implies that the previous state will be wiped out, so that might be worth clarifying in documentation somewhere.

Reply:
> Does this function have hysteresis, then? Where the previous state will have an effect on the current state? If so, it might be good to mention that in the documentation of updateOccupancy.

Hm, not sure if I understand your question correctly. The probability updates are done by Octomap (e.g., by `insertPointCloud`). Let me try to update documentation more correctly.

> But I think sometimes "update" implies that the previous state will be wiped out, so that might be worth clarifying in documentation somewhere.

I agree. The meaning of "update" term often cases too broad. I might use "set" in the case of wiping out the original value.